src/object/srv_obj.c

```diff
@@ -1726,8 +1724,17 @@ static void
 obj_ioc_end(struct obj_io_context *ioc, int err)
 {
 	if (ioc->ioc_began) {
+		struct obj_tls	*tls = obj_tls_get();
+		uint32_t	opc = ioc->ioc_opc;
+
 		dss_rpc_cntr_exit(DSS_RC_OBJ, !!err);
 		ioc->ioc_began = 0;
+
+		/** Update sensors */
+		/** XXX duration should probably be aborted in case of error */
+		d_tm_mark_duration_end(&tls->ot_op_lat[opc], NULL);
```

Comment:
if ot_op_lat/active/total is failed to be initialized in obj_tls_init(), probably we should just avoid these metrics? 
Instead if it keeps call these metrics api with null ops, it will try to recreate the metric path, is this what we want here?

Reply:
in this case, i will call d_tm_mark_duration_end(NULL,NULL). My suggestion is to change the function to make sure that they do nothing in this case. It will avoid adding if statement everywhere through the code.

Reply:
ah, ok, I had thought each d_tm_xx are suppose to handled NULL in all cases.

Reply:
If these functions are called with (NULL,NULL), they return -DER_INVAL.  If one of the args is valid, d_tm_decrement() and d_tm_incremement() will create the metric.  d_tm_mark_duration_end() is a little different, because it is paired with d_tm_mark_duration_start() which will do the creation.  The d_tm_mark_duration_end() will not create it under any circumstance.  That said, the way it is implemented here is correct.

Reply:
Ah, you are right. 
Hmm, it looks like duration_start() might mismatch duration_end(), if the current IO is yield.
for example this sequence
1. IO1 get RPC  ioc_begin() mark start time.  yield
2. IO2 get RPC ioc_begin() Update the start time? yield
3. IO1 complete RPC then get end time, but the latency will be current_time - IO2 start time?

Reply:
@jbrosenz probably we need multiple entries in node->dtn_metric->dtm_data  so we can remember all pairs of start/end ?
@johannlombardi  thoughts?

Reply:
Provided that tls->ot_op_lat[opc] is unique for IO1 and IO2, then in step 3 when IO1 completes the RPC the duration will be calculated as current_time - IO1 start time.  If you are saying that tls->ot_op_lat[opc] is used for IO1 and IO2, then you are right that the latency will be current_time - IO2 start time.  Are IO1 and IO2 sharing the same metric?

Reply:
If I understand correctly, these metric are per xstream.  yes, there could be multiple IO inflight at the same time within same xstream, since we do at least 2 yields (bulk & BIO) for 1 RPC handler/

Reply:
In that case, I recommend differentiating them so that a unique metric can be assigned to each.

Reply:
no, we can not afford one metric per RPC.

Reply:
Do I understand correctly that having one tls->ot_op_lat[opc] is valid, but having tls->ot_op_lat[<opc>_io1] and tls->ot_op_lat[<opc>_io2] is too many?  I'm not agreeing or disagreeing.  I'm just looking to understand the issue.

Reply:
oh, I mean assigning metric to each individual RPC (_io1, _io2...) will be way too much.
 
Though we probably have to pre-allocate enough lots to make sure all inflight IO will have its own start/end pair,  for example if there are 10 inflight IO, then it will need 10 pairs.

Unfortunately, there are no maxim limit for inflight I/O within one xstream, so it has to re-allocate the pairs if needed.   But we probably can set a hard limit here for example 32 or 64, and skip the latency counting if it is reach the limit. So api could be

d_tm_mark_duration_start(metric, clk_id, item, &index...)  //index = -1 if the pair is full, otherwise return the index of the pair.

d_tm_makr_duration_end(metric, item, index) // calculate the latency by index, then update the  tls->ot_op_lat[opc]

Reply:
all good points ... let me think about it

Reply:
My formatting in the comment didn't quite show what I wanted.  It was intended to ask if having one metric per opcode is significantly different than having two metrics per opcode.  It's obviously double the data .. but that is the point.  Is that an actual problem?

Reply:
The trick is that there is no real upper limit on concurrent I/Os we can handle.
In any case, i think that i have a solution in the latest version of the patch by using gauge instead.